### ISSUE TITLE
GRU4Rec: Session-parallel mini-batches

### ISSUE THREAD

**Issue by amoussawi** [mquad/sars_tutorial#5]
Hello! GRU4Rec (and HGRU4Rec) uses a session parallel mini-batches method to build mini-batches and train the model on them, where every mini-batch consists of the current event and the next event. The problem here is that the model weights are not updated after the whole session finishes, but at every one of those mini-batches (correct me if i'm wrong), and the hidden state computed using the previous model weights is kept to be used as input for the next mini-batch, isn't this practically wrong? is there a reason for this that i might be missing?
The reason I'm asking is that typical RNN models, such as the ones used for NLP, backpropagate after passing the whole word sequence.

And if this was intentional, have you found it to do better than passing the whole sequence?

**Comment by mquad** [comment]
Hi, correct, GRU4Rec and HGR4Rec don't use BBPT algorithm for training, but instead update the model weights at each step using traditional backprop on the "unrolled" sequence.

While this may actually generate the inconsistencies between hidden-states and model weights that you mention, we did not observe any particular disadvantage compared to training over the whole sequence, with the advantage of faster training times over really large datasets.

In any case, I'll be happy to hear your experience with that :-)

**Comment by amoussawi** [comment]
I'm actually also wondering whether the model is learning any meaningful hidden states, because the loss gradients are not being backpropagated back in time (right?), so the hidden state weights that were used to compute the input hidden state for the next event won't be updated. I must be missing something, so correct me please (and bear with me :p).

**Comment by amoussawi** [comment]
I'm actually also wondering whether the model is learning any meaningful hidden states, because the loss gradients are not being backpropagated back in time (right?), so the hidden state weights that were used to compute the input hidden state for the next event won't be updated. I must be missing something, so correct me please (and bear with me :p).

**Comment by amoussawi** [comment]
This worth experimenting against. Another difference is that loss gradients won't be backpropagated back in time, so hidden state weights won't be updated with future gradients, not sure how much this will make RNNs lose its temporal power. This actually avoids the whole vanishing gradient problem :p
